[Router] added appending of new optional document fragment

rodnaph opened the pull request:
added a new optional parameter to the generate method for the document fragment. when specified this is appended to generated urls.

| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | none |
| License | MIT |
| Doc PR | none |

rodnaph committed:
[Routing] adds _fragment special option to url generation for document fragment

Tobion commented:
I agree that a helper for the fragment is useful because of the encoding need. And I cannot think of a better solution than the one provided here. And the tiny BC break is negligible because it's unlikely.

Three things are missing for me to be merged:
1. the special meaning of `_fragment` must be documented in the phpdoc of UrlGeneratorInterface
2. the slash `/` should be decoded from the fragment just as it has been done for the query in #12223. this is also part of the rfc: "The characters slash ("/") and question mark ("?") are allowed to
   represent data within the fragment identifier."
3. we should add a test when`_fragment` is used as a path parameter name, e.g. `/fragment/{_fragment}` both for generating and matching.

fabpot commented:
@rodnaph Can you address @Tobion concerns?

rodnaph commented:
@Tobion 

I've addressed your first two concerns.

For the third I thought the best solution would be to throw a LogicException when compiling the route, as \_fragment has no meaning as a path parameter. I'm not sure what you meant about _"a test with a same document relative reference with just a fragment"_ - would that be a route with no pattern? eg. `new Route('');` ?

I also rebased onto current master as there were some failing lint issues.

/cc @fabpot

bendavies commented:
+1 @Tobion

stloyd commented on src/Symfony/Component/Routing/RouteCompiler.php:
```diff
@@ -59,6 +59,13 @@ public static function compile(Route $route)
         $pathVariables = $result['variables'];
+
+        foreach ($pathVariables as $pathParam) {
+            if ('_fragment' == $pathParam) {
```
Should be `===`.

rodnaph replied:
Though not documented, it looks like the `variables` output by compilePattern will always be strings, so a strict type equality would not be required.

Unless of course this is a coding standard, or maybe if a test could show it was required (as it seems to me without proof either way then both strict and non-strict comparison have the possibility of causing a bug).

stof replied:
Our coding standards are that we always use strict comparison unless we explicitly need the type juggling (because type juggling can actually cause totally unexpected behavior if you are not careful).
And given we know that strings are always returned, we know we don't need this type juggling.

stloyd commented on src/Symfony/Component/Routing/Generator/UrlGenerator.php:
```diff
@@ -272,14 +272,25 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
         }
 
+        if ($fragment) {
+            $validFragmentChars = array('%2F' => '/', '%3F' => '?');
```
No need for extra variable.

stloyd commented on src/Symfony/Component/Routing/RouteCompiler.php:
```diff
@@ -59,6 +59,13 @@ public static function compile(Route $route)
+
+        foreach ($pathVariables as $pathParam) {
+            if ('_fragment' == $pathParam) {
+                throw new \LogicException(sprintf('Route pattern "%s" cannot contain "_fragment" as a path parameter.', $route->getPath()));
```
`InvalidArgumentException` looks more proper here IMO.

inso commented on src/Symfony/Component/Routing/Generator/UrlGenerator.php:
```diff
@@ -272,14 +272,24 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
             $url .= '?'.strtr($query, array('%2F' => '/'));
         }
 
+        if ($fragment) {
```
this is broken for `$fragment = '0'`

rodnaph commented:
@inso Fixed, thanks.

Koc commented on src/Symfony/Component/Routing/RouteCompiler.php:
```diff
@@ -59,6 +60,13 @@ public static function compile(Route $route)
 
         $pathVariables = $result['variables'];
+
+        foreach ($pathVariables as $pathParam) {
```
`array_search`?

fabpot commented on src/Symfony/Component/Routing/Generator/UrlGenerator.php:
```diff
@@ -268,14 +268,24 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
         $extra = array_diff_key($parameters, $variables, $defaults);
+
+        // extract fragment
+        $fragment = isset($extra['_fragment']) ? $extra['_fragment'] : false;
```
the default value should be `""`, not `false`

fabpot commented:
Apart from my small comment, :+1: 

ping @Tobion 

@rodnaph Can you rebase on current master?

rodnaph commented:
@fabpot done

stloyd commented on src/Symfony/Component/Routing/Generator/UrlGenerator.php:
```diff
@@ -266,14 +266,24 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
             $url .= '?'.strtr($query, array('%2F' => '/'));
         }
 
+        if (strlen($fragment)) {
```
Should be `'' === $fragment`.

xabbuh replied:
`'' !== $fragment`

fabpot commented:
Thank you @rodnaph.

King2500 commented:
Why not use the more intuitive (and less conflicting) parameter key `#` instead of `_fragment`, like:

```php
// generating a URL with a fragment (/settings#password)
$this->get('router')->generate('user_settings', ['#' => 'password']);
```

javiereguiluz commented:
@King2500 I'm always in favor of short option names, but in this case I'd prefer to keep the consistency. Every special routing option in Symfony follows the same pattern: `_controller`, `_locale`, `_route`, `_route_params`, `_redirected`, etc.
